3.0 - Fixing existing translations

patrickconroy opened the pull request:
I missed one of these somehow. This fixes an issue with saving data with non-default locales.

Commit:
Fixing existing translations

I missed one of these somehow. This fixes an issue with saving data with non-default locales.

lorenzo commented on src/ORM/Behavior/TranslateBehavior.php:
```diff
@@ -482,8 +482,8 @@ protected function _bundleTranslatedFields($entity)
     {
-        $target = TableRegistry::get($this->_config['translationTable']);
-        $association = $this->_table->association($target->alias());
+        $targetAlias = Inflector::slug($this->_config['translationTable']);
```
Can you use `'_'` as the last argument for `slug()` here and in the Behavior?

markstory replied:
Or use `underscore()`.

dereuromark replied:
The latter better.
slug() should only be used when the input can be non alpha numeric as it does a lot of overhead.

patrickconroy replied:
Hm, I don't think underscore works, I tried changing it to that. My translationTables look like `Company/Plugin.I18n`, and so the associations break at some point since that underscores to `company/plugin.i18n`. I funneled all those `Inflector::slug` calls to `$this->foo` (`$targetAlias = $this->foo($table);`)

This works:
```
private function foo($table) {
    return Inflector::slug($table, '-');
}
```

This errors out with `Method Cake\ORM\Query::__toString() must not throw an exception`
```
private function foo($table) {
    return Inflector::underscore($table);
}
```

It's hard to tell where the error is happening (I think a recent core change is accidentally obfuscating errors, not sure yet).

markstory commented:
I hadn't considered the vendored plugin case. In that case slug() is better, but the replacement should be `_` not `-` which is the default.

patrickconroy commented:
Thanks @markstory, in that case the PR should be good to go.

patrickconroy committed:
Inflecting slug with dash instead of underscore

Commit:
Updating test for underscores instead of dashes

patrickconroy commented:
I think that failing check is unrelated, hopefully.

lorenzo commented:
@patrickconroy I was giving it another look and wondered why calculate a custom alias when we can just use `$table->alias()` after the `TableRegistry::get()`... is there any reason to not use that?

patrickconroy commented:
That works too but I wasn't sure if it'd be too much overhead to get the Table instance / alias when that value really just needs to represent a key as far as the Behavior's concerned. Let me know if you want me to change it to that.

lorenzo commented:
@patrickconroy getting the alias will definitely be faster than calculating a slug

patrickconroy commented:
Hm, ok what about saving that in the config in `__construct`, then? So it's called once instead of all over the place..
```
public function __construct(Table $table, array $config = [])
{
    $config += [
        'defaultLocale' => I18n::defaultLocale(),
        'translationTableInstance' => TableRegistry::get($config['translationTable'])
    ];
    parent::__construct($table, $config);
}
```

Or is that overkill? Maybe just a class var for that instead?

Commit:
Setting translationTable on the class to make it easier to work with

patrickconroy commented:
I made it a class var, that seemed to make sense, and the class is cleaner now.

Commit:
Fixing typo

Commit:
One more typo.

Commit:
Fixing a typo that I thought I fixed

Commit:
Fixing i18n when two Tables have the same name

Inflector::underscore is the only (Cake) way to ensure uniqueness in the TableRegistry.

patrickconroy commented:
Actually I just ran into an issue with this which is why I made it use slug.. if you have two I18n classes the TableRegistry needs their aliases to be unique. Since there's no uniqueness between these aliases, `$b` returns `$a`:
```
$a = TableRegistry::get('Company/News.I18n');
$b = TableRegistry::get('Company/Comments.I18n');
```

To solve that in the current state of the behavior you'd have to do this:

```
$translationAlias = Inflector::slug($this->_config['translationTable'], '_');

if (!TableRegistry::exists($translationAlias)) {
    $this->_translationTable = TableRegistry::get($translationAlias, [
        'className' => $this->_config['translationTable']
    ]);
} else {
    $this->_translationTable = TableRegistry::get($translationAlias);
}
```

Inflector::underscore won't work, since the TableRegistry will see the slashes and think it's a plugin - making the uniqueness still based on the final className.

lorenzo commented:
Thanks @patrickconroy
